### PR TITLE
Add fixes to spanish by ChanceNCounter

### PR DESCRIPTION
**Pull request by forslund**
This handles parsing cuarta among other smaller issues.

### CODE REVIEW

File: lingua_franca/lang/parse_es.py [hunk 1]
```diff
@@ -338,12 +339,12 @@ def clean_string(s):
             " ").replace(
             "_",
             "")
-        # handle synonims and equivalents, "tomorrow early = tomorrow morning
-        synonims = {"mañana": ["amanecer", "temprano", "muy temprano"],
+        # handle synonryms and equivalents, "tomorrow early = tomorrow morning
```
Comment (ChanceNCounter) [inline lingua_franca/lang/parse_es.py]: I seem to have replaced a typo in this comment with a different typo

File: test/test_parse_es.py [hunk 2]
```diff
@@ -84,6 +86,134 @@ def test_extract_number_es(self):
...
+        self.assertEqual(extract_number("dos e media", lang='es'), 2.5)
+        self.assertEqual(extract_number(
+            "catorce e milésima", lang='es'), 14.001)
+
+        self.assertEqual(extract_number("dos punto cero dos", lang='es'), 2.02)
+
+    def test_isFraction_es(self):
+        self.assertEqual(isFractional_es("vigésimo"), 1.0 / 20)
+        self.assertEqual(isFractional_es("trigésimo"), 1.0 / 30)
+        self.assertEqual(isFractional_es("centésima"), 1.0 / 100)
```
Comment (adocampo) [inline test/test_parse_es.py]: Also the male form can be said: "centésimo"

File: test/test_parse_es.py [hunk 3]
```diff
@@ -84,6 +86,134 @@ def test_extract_number_es(self):
...
         self.assertEqual(extract_number("seis punto dos", lang='es'), 6.2)
         self.assertEqual(extract_numbers("un medio", lang='es'), [0.5])
+        self.assertEqual(extract_number("cuarto", lang='es'), 0.25)
+
+        self.assertEqual(extract_number("2.0", lang='es'), 2.0)
+        self.assertEqual(extract_number("1/4", lang='es'), 0.25)
+
+        self.assertEqual(extract_number("dos e media", lang='es'), 2.5)
+        self.assertEqual(extract_number(
+            "catorce e milésima", lang='es'), 14.001)
```
Comment (adocampo) [inline test/test_parse_es.py]: In spanish we would say: "catorce coma cero cero uno" I hardly doubt someone says "catorce y una milésima"

File: lingua_franca/lang/parse_es.py (hunk 1 above)
Comment (forslund) [reply lingua_franca/lang/parse_es.py]: Thanks, updated the commit.

File: test/test_parse_es.py [hunk 4]
```diff
@@ -84,6 +86,134 @@ def test_extract_number_es(self):
...
+        self.assertEqual(extract_datetime(
+            "11 dic", lang='es', anchorDate=datetime(1998, 12, 1))[0],
+            datetime(1998, 12, 11))
+
+        self.assertEqual(extract_datetime("", lang='es'), None)
+
+    # TODO fix bug causing these tests to fail (MycroftAI/mycroft-core#2348)
+    #         reparar error de traducción preveniendo las funciones abajo de
+    #         retornar correctamente
+    #         (escribido con disculpas por un Inglés hablante)
```
Comment (adocampo) [inline test/test_parse_es.py]: escrito  ;)

File: test/test_parse_es.py [hunk 5]
```diff
@@ -84,6 +86,134 @@ def test_extract_number_es(self):
...
+        # It's also failing on years
+        self.assertEqual(extract_datetime(
+            "11 ago 1998", lang='es')[0], datetime(1998, 8, 11))
+
+    def test_extract_datetime_relative(self):
+        self.assertEqual(extract_datetime("esta noche", anchorDate=datetime(
+            1998, 1, 1), lang='es'),
+            [datetime(1998, 1, 1, 21, 0, 0), 'esta'])
+        self.assertEqual(extract_datetime(
+            "ayer noche", anchorDate=datetime(1998, 1, 1),
```
Comment (adocampo) [inline test/test_parse_es.py]: Yet "ayer por la noche" or "ayer noche" aren't wrong at all, we rather use "anoche" to specify "last night".

File: test/test_parse_es.py [hunk 6]
```diff
@@ -84,6 +86,134 @@ def test_extract_number_es(self):
...
+            lang='es')[0], datetime(1997, 12, 31, 21))
+        self.assertEqual(extract_datetime(
+            "el noche anteayer", anchorDate=datetime(1998, 1, 1),
```
Comment (adocampo) [inline test/test_parse_es.py]: We say "hace dos noches" or "anteanoche"

File: test/test_parse_es.py [hunk 7]
```diff
@@ -84,6 +86,134 @@ def test_extract_number_es(self):
...
+            lang='es')[0], datetime(1997, 12, 30, 21))
+        self.assertEqual(extract_datetime(
+            "el noche ante ante ayer", anchorDate=datetime(1998, 1, 1),
```
Comment (adocampo) [inline test/test_parse_es.py]: We say "hace tres noches" ("anteanteanoche" and "trasanteanoche" are no longer on the spanish dictionary since 2008 because no one say that way)

File: test/test_parse_es.py [hunk 8]
```diff
@@ -84,6 +86,134 @@ def test_extract_number_es(self):
...
+            lang='es')[0], datetime(1997, 12, 29, 21))
+        self.assertEqual(extract_datetime(
+            "mañana en la mañana", anchorDate=datetime(1998, 1, 1),
```
Comment (adocampo) [inline test/test_parse_es.py]: In proper spanish or es_ES, should be said "mañana **por** la mañana". _Some_ latam say "mañana **en** la mañana" but those usually are es_US or es_MX, more influenced by english language.

File: test/test_parse_es.py [hunk 9]
```diff
@@ -84,6 +86,134 @@ def test_extract_number_es(self):
...
+            lang='es')[0], datetime(1998, 1, 2, 8))
+        self.assertEqual(extract_datetime(
+            "ayer en la tarde", anchorDate=datetime(1998, 1, 1),
```
Comment (adocampo) [inline test/test_parse_es.py]: same than before. "Mañana **por** la tarde"

File: test/test_parse_es.py [hunk 10]
```diff
@@ -84,6 +86,134 @@ def test_extract_number_es(self):
...
+            lang='es')[0], datetime(1997, 12, 31, 15))
+
+        self.assertEqual(extract_datetime(
+            "que año es", anchorDate=datetime(1998, 1, 1),
```
Comment (adocampo) [inline test/test_parse_es.py]: Should it be accented? "qué año es" if it is a question, it should. (I don't know if that makes any sense of the unit test)

File: test/test_parse_es.py [hunk 11]
```diff
@@ -84,6 +86,134 @@ def test_extract_number_es(self):
...
+        self.assertEqual(extract_datetime("hoy 2 am", lang='es',
+                                          anchorDate=datetime(1998, 1, 1))[0],
+                         datetime(1998, 1, 1, 2))
+        self.assertEqual(extract_datetime("hoy 2 p.m.", lang='es',
+                                          anchorDate=datetime(1998, 1, 1))[0],
+                         datetime(1998, 1, 1, 14))
+        self.assertEqual(extract_datetime("hoy 2 pm", lang='es',
+                                          anchorDate=datetime(1998, 1, 1))[0],
+                         datetime(1998, 1, 1, 14))
+
```
Comment (adocampo) [inline test/test_parse_es.py]: In spanish we don't say "am" or "pm" but "de la madrugada", "de la mañana", "de la tarde", or "de la noche". The time frames varies from region to region, but usually are those:
- from 00:01 to 11:59 --> "de la mañana"
- from 12:01 to 20:00 --> "de la tarde"
- from 20:01 to 23:59 --> "de la noche"

There are some special time frames:
- 00:00 is "media noche" or "medianoche" and 12:00 is "medio día" or "mediodía"
- the time frame from 01:00 to 05:59 is usually also called "de la madrugada"

File: test/test_parse_es.py [hunk 12]
```diff
@@ -84,6 +86,134 @@ def test_extract_number_es(self):
...
+    @unittest.skip("These phrases are not parsing correctly.")
+    # parses as "morning" and returns 8:00 on anchorDate
+    def test_extract_datetime_relative_failing(self):
+        self.assertEqual(extract_datetime(
+            "mañana", anchorDate=datetime(1998, 1, 1), lang='es')[0],
+            datetime(1998, 1, 2))
```
Comment (adocampo) [inline test/test_parse_es.py]: When I tried to create parser_es.py I tried to differentiate "morning" from "tomorrow" by parsing the former word. In spanish "la mañana" is _always_ the time frame before midday, while the day after today is simply "mañana". That's the only way we could differentiate those two terms: if "mañana" is preceded by "la" is unmistakably "morning" and without is unmistakably "tomorrow", so even "mañana por la mañana" can be parsed correctly.

File: lingua_franca/lang/parse_es.py [hunk 13]
```diff
@@ -37,15 +37,14 @@ def isFractional_es(input_str):
     if input_str.endswith('s', -1):
         input_str = input_str[:len(input_str) - 1]  # e.g. "fifths"
 
-    aFrac = ["medio", "media", "tercio", "cuarto", "cuarta", "quinto",
-             "quinta", "sexto", "sexta", "séptimo", "séptima", "octavo",
-             "octava", "noveno", "novena", "décimo", "décima", "onceavo",
-             "onceava", "doceavo", "doceava"]
+    aFrac = {"medio": 2, "media": 2, "tercio": 3, "cuarto": 4,
+             "cuarta": 4, "quinto": 5, "quinta": 5, "sexto": 6, "sexta": 6,
+             u"séptimo": 7, u"séptima": 7, "octavo": 8, "octava": 8,
```
[the remark this answers is not in the document]
Comment (forslund) [reply lingua_franca/lang/parse_es.py]: Correct you are! Will remove those

File: test/test_parse_es.py [hunk 14]
```diff
@@ -84,6 +86,134 @@ def test_extract_number_es(self):
             lang='es')), [1, 4, 7, 8, 14, 157])
         self.assertEqual(extract_number("seis punto dos", lang='es'), 6.2)
         self.assertEqual(extract_numbers("un medio", lang='es'), [0.5])
+        self.assertEqual(extract_number("cuarto", lang='es'), 0.25)
+
+        self.assertEqual(extract_number("2.0", lang='es'), 2.0)
+        self.assertEqual(extract_number("1/4", lang='es'), 0.25)
+
+        self.assertEqual(extract_number("dos e media", lang='es'), 2.5)
```
[the remark this answers is not in the document]
Comment (forslund) [reply test/test_parse_es.py]: Ah humm, that makes the test fail...

[the remark this answers is not in the document]
Comment (forslund) [reply test/test_parse_es.py]: Ok think I fixed that...

[the remark this answers is not in the document]
Comment (adocampo) [reply test/test_parse_es.py]: oh, it can also be said "dos y medio", depending if the subject is male or female

[the remark this answers is not in the document]
Comment (adocampo) [reply test/test_parse_es.py]: or "dos coma cinco" | "dos punto cinco" if we are speaking of just decimal numbers

File: lingua_franca/lang/parse_es.py [hunk 15]
```diff
@@ -120,7 +120,7 @@ def extractnumber_es(text, short_scale=True, ordinals=False):
             break
 
         # number word and fraction
-        ands = ["e"]
+        ands = ["e", "y"]
```
Comment (adocampo) [inline lingua_franca/lang/parse_es.py]: "e" as number or fractional separator doesn't exist in spanish. That's copy pasted from PT or IT. 
"e" can be used as a "y" only and if only the next word begins with an "i", to avoid cacophonies.

Comment (forslund) [reply lingua_franca/lang/parse_es.py]: So it can't be used with any numbers?

Comment (adocampo) [reply lingua_franca/lang/parse_es.py]: there is no number starting with "i", so no.

Comment (forslund) [reply lingua_franca/lang/parse_es.py]: Thanks I'll remove the "e"

Comment (forslund) [reply lingua_franca/lang/parse_es.py]: Just for clarity: the test phrace "catorce e milésima" should be "catorce y milésima"

Comment (adocampo) [reply lingua_franca/lang/parse_es.py]: TL;DR: yes, but for that example in particular, I also commented this https://github.com/MycroftAI/lingua-franca/pull/62#discussion_r360832559.

But yes, you're right, "**and**" in spanish is "**y**". Always... on numbering. 
It is always "y" when naming numbers and fractions. "Y" is _always_ placed between tens and units. You have some exceptions, like numbers from 11 to 29

45: cuarenta y cinco 
1492: mil cuatrocientos noventa y dos
2.856.729: dos millones ochocientos cincuenta y seis mil setecientos veintinueve
11: once
25: veinticinco

You can find some exceptions finding an "e" when enumerating and the forthcoming word begins with an "i", for example: "Eres guapo e inteligente" (You are handsome and intelligent), as the phonema of "y" and "i" is the same, so we put an "e" to avoid the cacophony.

Returning to numbers, as no number begins with an "i" (uno, dos tres, cuatro, cinco, seis, siete, ocho, nueve, diez, cien, mil, millon, etc), you will never ever see an "e".

Comment (forslund) [reply lingua_franca/lang/parse_es.py]: Thanks for the explanation! :+1:

File: test/test_parse_es.py (hunk 12 above)
Comment (ChanceNCounter) [reply test/test_parse_es.py]: I thought this would be a trivial addition, add another condition there, but it turns out there are a number of places in the parser that require some tweaking to facilitate this. If one of you guys wants to take a stab, please do!

File: test/test_parse_es.py (hunk 4 above)
Comment (ChanceNCounter) [reply test/test_parse_es.py]: well, i guess that's the most accurate comment I've ever written, then!

(Ake: comment reads, roughly, "fix translation error preventing below functions from returning correctly. written with apologies by an English speaker." except 'escribido' should read 'escrito')

File: test/test_parse_es.py (hunk 5 above)
Comment (ChanceNCounter) [reply test/test_parse_es.py]: Sorry, I keep forgetting to check all of these. So far, this is the only one where I don't defer to you ;)

This particular test was to ensure that "ayer" and "noche" combined gets yesterday and evening, rather than returning either tonight or yesterday morning, as you'd get if you passed it "ayer" or "noche" separately.

That said, if there isn't one, there should be a test for "anoche" That's an oversight on my part.

File: test/test_parse_es.py (hunk 11 above)
Comment (forslund) [reply test/test_parse_es.py]: This has been updated now.

File: test/test_parse_es.py (hunk 5 above)
Comment (forslund) [reply test/test_parse_es.py]: Adding a test for _anoche_ (to work exactly like ayer noche) does not pass. Let's create an issue for this after the PR is merged.

**Comment by forslund** [comment]
I don't think I can handle all the requested changes. My suggestion is that we merge this in the current state so it's at the same state as the current core. and we add @adocampo as issues allowing people more skilled in Spanish to move those along.

If there's some changes that are too weird I should be able to remove those.

**Comment by adocampo** [comment]
I agree we could merge and start working piece by piece

**Comment by ChanceNCounter** [comment]
Currently working a little bit, since it hasn't been merged yet anyway, and there are plenty of changes listed here to be addressed.

File: test/test_parse_es.py [hunk 16]
```diff
@@ -84,6 +86,134 @@ def test_extract_number_es(self):
...
+        self.assertEqual(extract_number(
+            "catorce e milésima", lang='es'), 14.001)
+
+        self.assertEqual(extract_number("dos punto cero dos", lang='es'), 2.02)
```
[the remark this answers is not in the document]
Comment (ChanceNCounter) [reply test/test_parse_es.py]: The parser already handled both, but there weren't any tests for "coma." There are about to be.

File: test/test_parse_es.py (hunk 3 above)
Comment (ChanceNCounter) [reply test/test_parse_es.py]: Agreed. This test is just to make sure that it 1) catches milésima, and 2) correctly combines it with another number. You'd hardly ever say "14 and a thousandth" in English, either, but edge cases are half the point.

That's also what these 'e's were for - cover edge cases - but they're obviously superfluous now that 'e' has been removed. I'd never noticed there aren't any numbers ending in 'i'. I just dutifully made sure both ands were functional.

File: test/test_parse_es.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (ChanceNCounter) [reply test/test_parse_es.py]: I stand corrected, and it's a very good thing you noticed this, and yet another reminder why it's so important to test every possible category of input!

The word "coma" works, but the character `,` is not currently parsing.

[the remark this answers is not in the document]
Comment (ChanceNCounter) [reply test/test_parse_es.py]: Apparently nobody implemented logic for written decimals. It's just been passing the whole float through the whole parser. This is now its own issue, I reckon.

**Comment by forslund** [comment]
@ChanceNCounter do you mean you want us to hold off on merging?

**Comment by ChanceNCounter** [comment]
Yes, sorry, I'd have been clearer if I'd realized it was other-people-are-awake o'clock. I'll have this whittled down to "fixed stuff" and "stuff that needs its own PR" in a short while.

File: test/test_parse_es.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (adocampo) [reply test/test_parse_es.py]: Well, I guess we can live just with "punto" until someone can parse the comma. The point here is the human could say "dos coma cero dos" and Mycroft should understand it as "2.04". As Mycroft is a voice assistant, we won't type "2,02" in the near future.

PS: I'm wondering how french or german people deals with this, as afaik, in those languages decimal separator is comma as well.

File: test/test_parse_es.py (hunk 12 above)
Comment (ChanceNCounter) [reply test/test_parse_es.py]: relevant to #64

@adocampo apart from "esta mañana," are there any other exceptions that we'll need to account for?

File: test/test_parse_es.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (ChanceNCounter) [reply test/test_parse_es.py]: Yeah, I've put up an issue to add that logic. Happy to tackle it myself, as well, but I've got appointments through the middle of the day, so I won't get to it until this evening at the earliest.

At its most basic, it's just a regex search for a whole word consisting entirely of numbers separated by a comma.

That won't work in the very long term, because somebody might type in `2,02E`, but it would at least solve `2,02`.

I agree that it's a low priority, as the module is mainly for Mycroft and Mycroft is a voice assistant, but it's entirely conceivable that somebody could - right this minute! - try to parse a bunch of text for Mycroft to pronounce, and it would fail on commas-as-decimals.

File: test/test_parse_es.py (hunk 12 above)
Comment (adocampo) [reply test/test_parse_es.py]: When you say other exceptions you mean all combinations of  "_whatever_ mañana" meaning "morning" don't you? For example
"esta mañana" means "this morning"
"hoy por/en la mañana" means "today in the morning"
"mañana por/en la mañana" means "tomorrow morning"

So, no. I think generally we can say "la mañana" for the time frame between midnight and noon with the sole exception of "esta mañana" which means "this morning".

Good point!

**Comment by forslund** [comment]
Thanks @ChanceNCounter, merging now.
